src/print.S

```diff
@@ -62,57 +270,506 @@ print_int:
+    retq
+
+/* ======================================================
+ * Find the shortest decimal representation within
+ * the interval of possible outputs.
+ * General procedure:
+ * Divide the whole output interval by 10 for as long as
+ * the upper boundary is greater as the lower boundary.
+ * Count the amount of removed digits and get the correct exponent
+ * with a call to calculate_final_exponent.
+ * The mantissa is calculated by taking the middle of the interval,
+ * as soon as the loop condition is false and adding 1, whenever
+ * the remainder of the previous division is greater than 5 or
+ * when the middle of the interval is equal to the bottom of the interval.
+ *
+ * NOTE: This is the simple implementation, which fits 96% of the cases
+ * according to the algorithm author. Which means that this algorithm fails
+ * to compute an accurate result for ~4% of the numbers.
+ *
+ * parameter:
+ *  rdi => pointer to structure:
+ *      +0  => sign bit
+ *      +4  => exponent
+ *      +8  => mantissa
+ *      +12 => adjusted exponent
+ *      +16 => adjusted mantissa
+ *      +20 => lower bound output interval
+ *      +24 => middle of output interval
+ *      +28 => upper bound output interval
+ *      +32 => lower bound interval conversion
+ *      +36 => middle of interval conversion
+ *      +40 => upper bound of interval conversion
+ *
+ * return:
+ *  eax => result mantissa
+ *  edx => result exponent
+ */
+find_shortest_representation:
+    pushq %rbp
+    movq %rsp, %rbp
+    pushq %rbx
+    pushq %r12 /* amount of removed digits */
+
+    movl 32(%rdi), %r8d
+    movl 36(%rdi), %r9d
+    movl 40(%rdi), %r10d
+    xorq %r12, %r12
+.Lfind_representation_loop:
+    movl $10, %ebx
+    movl %r8d, %eax
+    cqto
+    idivl %ebx
+    movl %eax, %r8d
+    movl %r9d, %eax
+    cqto
+    idivl %ebx
+    movl %eax, %r9d
+    movb %dl, %r11b
+    movl %r10d, %eax
+    cqto
+    idivl %ebx
+    movl %eax, %r10d
+    incl %r12d
+    movl %r10d, %eax
+    cqto
+    idivl %ebx
+    movl %eax, %ecx
+    movl %r8d, %eax
+    cqto
+    idivl %ebx
+    subl %eax, %ecx
+    cmpl $0, %ecx
+    jg .Lfind_representation_loop
+    pushq %rdi
+    movl 12(%rdi), %edi
+    movl %r12d, %esi
+    callq calculate_final_exponent
+    popq %rdi
+    movl %eax, %edx
+    movl %r9d, %eax
+    cmpl %r9d, %r8d
+    je .Lfind_representation_increase_output
+    cmpl $5, %r11d
+    jge .Lfind_representation_increase_output
+    jmp .Lfind_representation_return
+.Lfind_representation_increase_output:
+    incl %eax
+.Lfind_representation_return:
+    popq %r12
+    popq %rbx
+    movq %rbp, %rsp
+    popq %rbp
+    retq
+
+/* ======================================================
+ * convert a IEEE754 floating-point number to a string and
+ * print each character.
+ *
+ * parameter:
+ *  xmm0 => input floating-point number
+ */
+print_float:
+    pushq %rbp
+    movq %rsp, %rbp
+    pushq %r12 /* sign bit */
+    leaq -160(%rbp), %rdi
+    callq split_into_parts
```

Comment:
(Note that if it is 160 bytes long, the original code had an other error: that the saved value of %r12 in `pushq %r12` was part of the struct, because you were counting 160 off %rbp, and not %rsp. But it was fine if the struct was smaller.)